zaza/openstack/charm_tests/openidc/tests.py

```diff
@@ -0,0 +1,164 @@
+from zaza.openstack.charm_tests.neutron.setup import (
+    OVERCLOUD_NETWORK_CONFIG,
+    DEFAULT_UNDERCLOUD_NETWORK_CONFIG,
+)
+from zaza.openstack.charm_tests.nova.setup import manage_ssh_key
+from zaza.openstack.charm_tests.openidc.setup import (
+    FEDERATED_DOMAIN,
+    IDP,
+    PROTOCOL_NAME,
+)
+from zaza.openstack.utilities import (
+    generic as generic_utils,
+    openstack as openstack_utils,
+)
+
+# static users created by openidc-test-fixture charm
+OIDC_TEST_USER = 'johndoe'
+OIDC_TEST_USER_PASSWORD = 'f00bar'
+
+
+class CharmKeystoneOpenIDCTest(BaseKeystoneTest, test_utils.OpenStackBaseTest):
+    """Charm Keystone OpenID Connect tests."""
+
+    run_resource_cleanup = True
+    RESOURCE_PREFIX = 'zaza-openidc'
+
+    def setUp(self):
+        """Define openrc credentials for OIDC_TEST_USER."""
+        super().setUp()
+        charm_config = zaza.model.get_application_config('keystone-openidc')
+        client_id = charm_config['oidc-client-id']['value']
+        client_secret = charm_config['oidc-client-secret']['value']
+        metadata_url = charm_config['oidc-provider-metadata-url']['value']
+        self.oidc_test_openrc = {
+            'API_VERSION': 3,
+            'OS_USERNAME': OIDC_TEST_USER,
+            'OS_PASSWORD': OIDC_TEST_USER_PASSWORD,
+            # using the first keystone ip by default, for environments with
+            # HA+TLS enabled this is the virtual IP, otherwise it will be one
+            # of the keystone units.
+            'OS_AUTH_URL': 'https://{}:5000/v3'.format(self.keystone_ips[0]),
+            'OS_PROJECT_DOMAIN_NAME': FEDERATED_DOMAIN,
+            'OS_PROJECT_NAME': '{}_project'.format(OIDC_TEST_USER),
+            'OS_CACERT': openstack_utils.get_cacert(),
+            # openid specific info
+            'OS_AUTH_TYPE': 'v3oidcpassword',
+            'OS_DISCOVERY_ENDPOINT': metadata_url,
+            'OS_OPENID_SCOPE': 'openid email profile',
+            'OS_CLIENT_ID': client_id,
+            'OS_CLIENT_SECRET': client_secret,
+            'OS_IDENTITY_PROVIDER': IDP,
+            'OS_PROTOCOL': PROTOCOL_NAME,
+        }
+        logging.info('openrc: %s', pprint.pformat(self.oidc_test_openrc))
+
+    def test_token_issue(self):
+        """Test token issue with a federated user via openidc."""
+        openrc = copy.deepcopy(self.oidc_test_openrc)
+        with self.v3_keystone_preferred():
+            for ip in self.keystone_ips:
+                logging.info('keystone IP %s', ip)
+                openrc['AUTH_URL'] = 'https://{}:5000/v3'.format(ip)
+                keystone_session = openstack_utils.get_keystone_session(
+                    openrc, scope='PROJECT')
+                logging.info('Retrieving token for federated user')
+                token = keystone_session.get_token()
+                logging.info('Token: %s', token)
+                self.assertIsNotNone(token)
+        logging.info('OK')
+
+    def test_10_network_configuration(self):
```

Comment:
Is this a set-up function?  it "looks" like a test, but seems to just configure things?